File: Piano/KeyboardLightController.cpp
// Copyright(c) Microsoft Open Technologies, Inc.All rights reserved.Licensed under the BSD 2 - Clause License.See License.txt in the project root for license information.
//

#include "KeyboardLightController.h"

KeyboardLightController::KeyboardLightController(
    const unsigned char SLAVE_SELECT_PIN_
    ) :
    _SLAVE_SELECT_PIN(SLAVE_SELECT_PIN_)
{}

// sets up the SPI for the Keyboard Lighting
void
KeyboardLightController::begin(
void
) {
    // Initialize SPI communication to MCP23S17
    SPI_Missing::begin(_SLAVE_SELECT_PIN);

    // Set GPA to OUTPUT
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, CTLREG_BK0_IODIRA, SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, 0, SPI_LAST);

    // Set GPB to OUTPUT
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, CTLREG_BK0_IODIRB, SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, 0, SPI_LAST);
}

// this will turn off the lights set by the bitmask
unsigned short
KeyboardLightController::disableLights(
const unsigned short BITMASK_
) {
    return set(BITMASK_, 0);
}

// this will turn on the lights set by the bitmask
unsigned short
KeyboardLightController::enableLights(
const unsigned short BITMASK_
) {
    return set(BITMASK_, 1);
}

// To be called when ending the SPI communication with the keyboard lights.
void
KeyboardLightController::end(
void
) {
    // Set GPA to INPUT
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, CTLREG_BK0_IODIRA, SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, 1, SPI_LAST);

    // Set GPB to INPUT
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, CTLREG_BK0_IODIRB, SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, 1, SPI_LAST);

    // Finalize SPI communication to MCP23S17
    SPI_Missing::end(_SLAVE_SELECT_PIN);
}

// Alternates flashing of the keyboard lights 0 and 15 to provide a heartbeat
void
KeyboardLightController::heartbeat(
void
) {
    static unsigned char bank_a = 1;

    // Transmit bytes
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, CTLREG_BK0_GPIOA, SPI_CONTINUE);

    if (bank_a) {
        SPI_Missing::transfer(_SLAVE_SELECT_PIN, 0x01, SPI_CONTINUE);
        SPI_Missing::transfer(_SLAVE_SELECT_PIN, 0x00, SPI_LAST);
    }
    else {
        SPI_Missing::transfer(_SLAVE_SELECT_PIN, 0x00, SPI_CONTINUE);
        SPI_Missing::transfer(_SLAVE_SELECT_PIN, 0x80, SPI_LAST);
    }

    // Toggle bank
    bank_a = !bank_a;
}

// Sets the lights given by the bitmask based on the controller (CTL_).
// Note: does not affect pins 0 and 15
unsigned short
KeyboardLightController::set(
const unsigned short BITMASK_,
const unsigned char CTL_
) {
    const unsigned char *gp = 0;

    // Apply bitmask relative or absolute
    switch (CTL_) {
    case 0:
        // Takes a flipped version of the bitmask. Then ANDs it with the current state
        _state &= ~BITMASK_;
        break;
    case 1:
        // Takes the bitmask and ORs it with the state
        _state |= BITMASK_;
        break;
    default:
        // state is equal to the bitmask
        _state = BITMASK_;
    }

    // Shift bits for pin mapping (0 and 15 are used for the heartbeat)
    // Note: Take this out if you want to use all the pins
    _state <<= 2;
    _state >>= 1;

    // Break into two bytes for transmission
    gp = (const unsigned char *) &_state;

    // Transmit bytes
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, CTLREG_BK0_GPIOA, SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, gp[0], SPI_CONTINUE);
    SPI_Missing::transfer(_SLAVE_SELECT_PIN, gp[1], SPI_LAST);

    // Shift bits for storage
    _state >>= 1;

    return _state;
}

// Sets the lights on and off as designated by the bitmask
unsigned short
KeyboardLightController::setLights(
const unsigned short BITMASK_
) {
    return set(BITMASK_, 0xFF);
}

File: build2014-piano/KeyboardLightController.cs
using System;
using System.Threading.Tasks;
using Windows.Devices.Spi;

/*
KeyboardLightController Static Library

This library will control keyboard lighting
using pins 1-14 (ignoring 0 and 15) of the
MCP23S17 GPIO port expander via SPI.

SPI requires the following pins:
SS - Slave select
MOSI - Master out Slave in
SCK - Serial Clock

The circuit:
MCP23S17 - GPIO (general purpose input/output) port expander
14 N-channel MOSFETs - One for each LED, triggering 12V/1A of current.
LED Strip - RadioShack LED Waterproof Flexi Strip 60 LED 1M (White)
12V/5A power supply
*/

namespace build2014_piano
{
    class KeyboardLightController
    {
        public enum SetAction
        {
            Disable,
            Enable,
            Force
        }

        //%%%%%%%%%%%/ MCP23S17 Op-codes /%%%%%%%%%%%//
        const byte SPI_SLAVE_ID = 0x40;    //0b01000000
        const byte SPI_SLAVE_ADDR = 0x00;  //0b00000000
        const byte SPI_SLAVE_WRITE = 0x00; //0b00000000

        //%%/ MCP23S17 Control Register Addresses /%%//
        const byte CTLREG_IODIRA = 0x00;   //0b00000000
        const byte CTLREG_IODIRB = 0x01;   //0b00000001
        const byte CTLREG_GPIOA = 0x12;    //0b00010010
        const byte CTLREG_GPIOB = 0x13;    //0b00010011

        public KeyboardLightController(SpiDevice spi_) {
            _spi = spi_;
        }

        /// <summary>
        /// Sets up the MCP23S17 for the Keyboard Lighting
        /// </summary>
        public void Begin() {
            // Set GPIO pins to OUTPUT
            _spi?.Write(new byte[] { (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), CTLREG_IODIRA, 0x00, 0x00 });
        }

        /// <summary>
        /// This will turn off the lights indicated by the bitmask
        /// </summary>
        /// <param name="bitmask_"></param>
        /// <returns></returns>
        public ushort DisableLights(ushort bitmask_) {
            return Set(bitmask_, SetAction.Disable);
        }

        /// <summary>
        /// This will turn on the lights indicated by the bitmask
        /// </summary>
        /// <param name="bitmask_"></param>
        /// <returns></returns>
        public ushort EnableLights(ushort bitmask_) {
            return Set(bitmask_, SetAction.Enable);
        }

        /// <summary>
        /// To be called when ending the SPI communication with the keyboard lights
        /// </summary>
        public void End () {
            // Set GPIO pins to INPUT
            _spi?.Write(new byte[] { (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), CTLREG_IODIRA, 0xFF, 0xFF });
        }

        /// <summary>
        /// Alternates flashing of the status lights on 0 and 15 to provide a heartbeat
        /// </summary>
        public void Heartbeat () {
            byte hi = 0x00;
            byte lo = 0x00;

            // Set heartbeat
            if (_blink_bank_a)
            {
                lo = 0x01;
            }
            else
            {
                hi = 0x80;
            }
            _blink_bank_a = !_blink_bank_a;

            TransmitStateUsingHiLoHeartbeatMask(hi, lo);
        }

        /// <summary>
        /// Sets the lights on and off as designated by the bitmask
        /// </summary>
        /// <param name="bitmask_"></param>
        /// <returns></returns>
        public ushort SetLights(ushort bitmask_) {
            return Set(bitmask_, SetAction.Force);
        }

        private bool _blink_bank_a = false;
        private object _mutex = new object();
        private SpiDevice _spi = null;
        private ushort _state = 0x0000;

        /// <summary>
        /// Sets the lights given by the bitmask based on the action parameter (SetAction)
        /// </summary>
        /// <param name="bitmask_"></param>
        /// <param name="action_"></param>
        /// <returns></returns>
        /// <remarks>
        /// Note: Does not affect pins 0 and 15
        /// </remarks>
        private ushort Set (ushort bitmask_, SetAction action_) {
            // Store heartbeat mask
            byte hi = (byte)((_state >> 8) & 0x80);
            byte lo = (byte)(_state & 0x01);

            lock (_mutex)
            {
                // Apply bitmask to state
                switch (action_)
                {
                    case SetAction.Disable:
                        _state &= (ushort)(~bitmask_);
                        break;
                    case SetAction.Enable:
                        _state |= bitmask_;
                        break;
                    case SetAction.Force:
                        _state = bitmask_;
                        break;
                    default:
                        break;
                }
            }

            return TransmitStateUsingHiLoHeartbeatMask(hi, lo);
        }


        /// <summary>
        /// Rolls a light bar (five wide) across the pins
        /// </summary>
        async public
        void Shimmer()
        {
            ushort key_mask = 0x0000;
            for (int i = -4; i < 15; ++i)
            {
                ushort light_bar = 0x003E;
                ushort user_mask = (ushort)(~key_mask & _state);
                int shift_count = Math.Abs(i);
                if (i < 0)
                {
                    key_mask = (ushort)(light_bar >> shift_count);
                }
                else
                {
                    key_mask = (ushort)(light_bar << shift_count);
                }
                SetLights((ushort)(key_mask | user_mask));

                await Task.Delay(100);
            }
        }

        /// <summary>
        /// Applies heartbeat to register values and transmits on the spi bus
        /// </summary>
        /// <param name="hi_"></param>
        /// <param name="lo_"></param>
        /// <returns></returns>
        private ushort TransmitStateUsingHiLoHeartbeatMask (byte hi_, byte lo_)
        {
            // Ensure valid heartbeat mask
            byte hi = (byte)(hi_ & 0x80);
            byte lo = (byte)(lo_ & 0x01);

            // Break into two bytes for transmission and apply heartbeat mask
            hi |= (byte)(_state >> 8);
            lo |= (byte)(_state);

            // Reset heartbeat
            _state >>= 1;
            _state <<= 2;
            _state >>= 1;

            // Transmit bytes
            _spi?.Write(new byte[] { (SPI_SLAVE_ID | SPI_SLAVE_ADDR | SPI_SLAVE_WRITE), CTLREG_GPIOA, lo, hi });

            return _state;
        }
    }
}
